File: sdks/java/src/test/java/com/swissarmynoife/sdk/SakClientTest.java
package com.swissarmynoife.sdk;

import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class SakClientTest {

  @Test
  void stripsTrailingSlash() {
    SakClient c = new SakClient("http://127.0.0.1:8787/");
    assertEquals("http://127.0.0.1:8787", c.baseUrl());
  }

  @Test
  void health() throws Exception {
    try (LocalJsonServer srv = LocalJsonServer.start("/health", "{\"ok\":true}")) {
      SakClient c = new SakClient(srv.baseUrl());
      assertEquals(Boolean.TRUE, c.health().get("ok"));
    }
  }

  @Test
  void listHelpers() throws Exception {
    record Case(String path, String body, ThrowingFn fn) {}
    Case[] cases =
        new Case[] {
          new Case("/v1/sak/modules", "{\"modules\":[]}", SakClient::listModules),
          new Case("/v1/sak/compute/work", "{\"work\":[]}", SakClient::listWork),
          new Case("/v1/sak/compute/nodes", "{\"nodes\":[]}", SakClient::listNodes),
          new Case(
              "/v1/sak/capacity",
              "{\"snapshot\":{\"total_ram_mb\":1}}",
              SakClient::capacity),
        };
    for (Case tc : cases) {
      try (LocalJsonServer srv = LocalJsonServer.start(tc.path, tc.body)) {
        SakClient c = new SakClient(srv.baseUrl());
        Map<String, Object> out = tc.fn.apply(c);
        assertTrue(out != null && !out.isEmpty());
      }
    }
  }

  @Test
  void enqueueWork() throws Exception {
    AtomicReference<String> posted = new AtomicReference<>();
    try (LocalJsonServer srv =
        LocalJsonServer.startPost(
            "/v1/sak/compute/work",
            posted,
            "{\"action\":\"enqueue\",\"work\":{\"status\":\"queued\"}}")) {
      SakClient c = new SakClient(srv.baseUrl());
      Map<String, Object> out = c.enqueueWork("echo", Map.of("n", 1));
      assertEquals("enqueue", out.get("action"));
      assertTrue(posted.get().contains("\"action\":\"enqueue\""));
      assertTrue(posted.get().contains("\"kind\":\"echo\""));
    }
  }

  @Test
  void requeueAndClaim() throws Exception {
    AtomicReference<String> posted = new AtomicReference<>();
    try (LocalJsonServer srv =
        LocalJsonServer.startPost(
            "/v1/sak/compute/work",
            posted,
            "{\"action\":\"ok\",\"work\":{\"id\":\"w1\"}}")) {
      SakClient c = new SakClient(srv.baseUrl());
      c.requeueWork("w1");
      assertTrue(posted.get().contains("requeue"));
      c.claimWork("n1");
      assertTrue(posted.get().contains("claim"));
      c.completeWork("w1", "n1", null);
      assertTrue(posted.get().contains("complete"));
      c.getWork("w1");
      assertTrue(posted.get().contains("\"action\":\"get\""));
    }
  }

  @FunctionalInterface
  interface ThrowingFn {
    Map<String, Object> apply(SakClient c) throws Exception;
  }

  /** Minimal JSON stub server. */
  static final class LocalJsonServer implements AutoCloseable {
    private final HttpServer server;
    private final int port;

    private LocalJsonServer(HttpServer server, int port) {
      this.server = server;
      this.port = port;
    }

    String baseUrl() {
      return "http://127.0.0.1:" + port;
    }

    static LocalJsonServer start(String path, String json) throws IOException {
      HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
      server.createContext(
          path,
          exchange -> {
            byte[] bytes = json.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, bytes.length);
            try (OutputStream os = exchange.getResponseBody()) {
              os.write(bytes);
            }
          });
      server.start();
      return new LocalJsonServer(server, server.getAddress().getPort());
    }

    static LocalJsonServer startPost(String path, AtomicReference<String> bodyOut, String json)
        throws IOException {
      HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
      server.createContext(
          path,
          exchange -> {
            bodyOut.set(new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8));
            byte[] bytes = json.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, bytes.length);
            try (OutputStream os = exchange.getResponseBody()) {
              os.write(bytes);
            }
          });
      server.start();
      return new LocalJsonServer(server, server.getAddress().getPort());
    }

    @Override
    public void close() {
      server.stop(0);
    }
  }
}

File: sdks/kotlin/src/test/kotlin/com/swissarmynoife/sdk/SakClientTest.kt
package com.swissarmynoife.sdk

import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import java.util.concurrent.atomic.AtomicReference
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SakClientTest {
    @Test
    fun stripsTrailingSlash() {
        val c = SakClient("http://127.0.0.1:8787/")
        assertEquals("http://127.0.0.1:8787", c.baseUrl)
    }

    @Test
    fun health() {
        LocalJsonServer.start("/health", """{"ok":true}""").use { srv ->
            val out = SakClient(srv.baseUrl).health()
            assertEquals(true, out["ok"])
        }
    }

    @Test
    fun listHelpers() {
        data class Case(val path: String, val body: String, val call: (SakClient) -> Map<String, Any?>)
        val cases =
            listOf(
                Case("/v1/sak/modules", """{"modules":[]}""", SakClient::listModules),
                Case("/v1/sak/compute/work", """{"work":[]}""", SakClient::listWork),
                Case("/v1/sak/compute/nodes", """{"nodes":[]}""", SakClient::listNodes),
                Case("/v1/sak/capacity", """{"snapshot":{"total_ram_mb":1}}""", SakClient::capacity),
            )
        for (tc in cases) {
            LocalJsonServer.start(tc.path, tc.body).use { srv ->
                val out = tc.call(SakClient(srv.baseUrl))
                assertTrue(out.isNotEmpty())
            }
        }
    }

    @Test
    fun enqueueWork() {
        val posted = AtomicReference<String>()
        LocalJsonServer.startPost(
            "/v1/sak/compute/work",
            posted,
            """{"action":"enqueue","work":{"status":"queued"}}""",
        ).use { srv ->
            val out = SakClient(srv.baseUrl).enqueueWork("echo", mapOf("n" to 1))
            assertEquals("enqueue", out["action"])
            assertTrue(posted.get().contains(""""action":"enqueue""""))
            assertTrue(posted.get().contains(""""kind":"echo""""))
        }
    }

    @Test
    fun requeueAndClaim() {
        val posted = AtomicReference<String>()
        LocalJsonServer.startPost(
            "/v1/sak/compute/work",
            posted,
            """{"action":"ok","work":{"id":"w1"}}""",
        ).use { srv ->
            val c = SakClient(srv.baseUrl)
            c.requeueWork("w1")
            assertTrue(posted.get().contains("requeue"))
            c.claimWork("n1")
            assertTrue(posted.get().contains("claim"))
            c.completeWork("w1", "n1")
            assertTrue(posted.get().contains("complete"))
            c.getWork("w1")
            assertTrue(posted.get().contains(""""action":"get""""))
        }
    }

    private class LocalJsonServer(
        private val server: HttpServer,
        val baseUrl: String,
    ) : AutoCloseable {
        override fun close() {
            server.stop(0)
        }

        companion object {
            fun start(path: String, json: String): LocalJsonServer {
                val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
                server.createContext(path) { exchange ->
                    val bytes = json.toByteArray(StandardCharsets.UTF_8)
                    exchange.responseHeaders.add("Content-Type", "application/json")
                    exchange.sendResponseHeaders(200, bytes.size.toLong())
                    exchange.responseBody.use { it.write(bytes) }
                }
                server.start()
                return LocalJsonServer(server, "http://127.0.0.1:${server.address.port}")
            }

            fun startPost(
                path: String,
                bodyOut: AtomicReference<String>,
                json: String,
            ): LocalJsonServer {
                val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
                server.createContext(path) { exchange ->
                    bodyOut.set(String(exchange.requestBody.readBytes(), StandardCharsets.UTF_8))
                    val bytes = json.toByteArray(StandardCharsets.UTF_8)
                    exchange.responseHeaders.add("Content-Type", "application/json")
                    exchange.sendResponseHeaders(200, bytes.size.toLong())
                    exchange.responseBody.use { it.write(bytes) }
                }
                server.start()
                return LocalJsonServer(server, "http://127.0.0.1:${server.address.port}")
            }
        }
    }
}